### PR TITLE
PDF Formatting related to references and bibliography

### PR DESCRIPTION
**Pull request by naomatheus**
#### Summary

This PR addresses:

- Bibliography (References) section and ensures that a bibliography is added to the document
- Allow inline references and link to the bibliography, expected behavior
- Text can be underlined in rich text editor and will appear in PDF
- Support for multiple inline references in a row (PR will be updated with this functionality)
~~- Line breaks will appear in PDF between paragraph elements and non-list elements~~
~~- *Note:* There is an extra section called References that is flagged for removal after some testing. It will leave a comment in the document which reads "pending removal." Please ignore during initial review.~~

PR does not yet include:

- Support for line breaks (this broke other formatting in certain situations)

To review:
- Create a simple document with images, tables, and equations; such as the attached example
[test-formatting-v1-0-journal (22).pdf](https://github.com/NASA-IMPACT/nasa-apt/files/10347912/test-formatting-v1-0-journal.22.pdf)

### CODE REVIEW

**Comment by sunu** [comment]
Hi @naomatheus! I am in the process of setting up APT locally, and once that is complete, I will be happy to provide a more detailed review.

In the meantime, I did notice a lot of commented out lines in the diff. I understand that these may be useful for future reference, but as someone new to the codebase, I think it would be helpful to remove them or stash them using `git stash` to make the review process easier. What do you think?

**Comment by naomatheus** [comment]
Hey @sunu. I saw your feedback for this commit. I've actually realized that there are some changes I made that inadvertently are affecting the loading of images and tables in certain situations. I'm going to pull only certain changes from this commit thread and I will open a PR which does not have side effects. 
Thanks for looking at this. I expect to close this PR tomorrow or next day.

**Comment by naomatheus** [comment]
Hey @sunu  I have force updated this PR and now it is ready for review of the functionality. 
And I also tried to remove a lot of the codebase that is commented.
Thanks for your feedback. Let me know if there's the same problem with this branch now and I can clean it up even further.

**Comment by naomatheus** [comment]
@thenav56  I have added one more commit to this PR. 
Could you verify that the issue #606  is resolved by this PR as well. 

cc: @wrynearson

**Comment by wrynearson** [comment]
Thanks @naomatheus, I think @thenav56 will be able to review this early this week. I will try to get APT set up locally today so I can test functionality as well.

File: app/pdf/generator.py [hunk 1]
```diff
@@ -699,16 +701,14 @@ def generate_latex(atbd: Atbds, filepath: str, journal=False):  # noqa: C901
             doc.append(Command("item", arguments=keypoint.strip("-") or " "))
         doc.append(Command("end", arguments="keypoints"))
 
-    print("CALLING GENERATE BIB FILE")
+    print(f"CALLING GENERATE BIB FILE at {filepath}.bib")
```
Comment (thenav56) [inline app/pdf/generator.py:704]: Debug code

File: app/pdf/generator.py [hunk 2]
```diff
@@ -757,7 +757,12 @@ def generate_latex(atbd: Atbds, filepath: str, journal=False):  # noqa: C901
         for _indx, item in enumerate(value["contents"]):
 
             # append item in order
-            doc.append(item)
+            if type(item) == NoEscape:
+
+                doc.extend([item])
+
+            else:
+                doc.append(item)
```
Comment (thenav56) [inline app/pdf/generator.py:765]: They are same. It should be okay to use append only.

File: app/pdf/generator.py [hunk 3]
```diff
@@ -849,113 +847,11 @@ def generate_latex(atbd: Atbds, filepath: str, journal=False):  # noqa: C901
...
-            continue
-
-        # SECTION CONTACT
-        if section_name == "contacts":
-
-            pass
-
-        if not document_data.get(section_name):
-            doc.append(process(CONTENT_UNAVAILABLE))  # type: ignore
-            continue
```
Comment (thenav56) [inline app/pdf/generator.py:910]: Just flagging if this is expected. My hunch is this is a duplicate, seeing the same code on the other PR.

File: app/pdf_utils/build_references.py [hunk 4]
```diff
@@ -1,33 +1,57 @@
...
+
+
+def create_reference(reference_id: str) -> NoEscape:
+    """
+    Returns a Latex formatted refrerence object, wrapped with a
+    NoEscape command to ensure the string gets processed as a
+    command and not plain text.
+    """
+
+    return NoEscape(f"\\cite{{{bib_reference_name(reference_id)}}}")
```
Comment (thenav56) [inline app/pdf_utils/build_references.py:24]: Debug code

File: app/pdf_utils/build_references.py [hunk 5]
```diff
@@ -1,33 +1,57 @@
...
-    except Exception as e:
-        raise e
+
+    reference_id = bib_reference_name(reference_dict["id"])
+    reference_id = reference_dict["id"]
+    print(reference_id)  # TODO remove this. Only to resolve precommit errors
+
+    # remove the id
+    del reference_dict["id"]
+    reference = ""
```
Comment (thenav56) [inline app/pdf_utils/build_references.py:38]: Debug code

File: app/pdf_utils/fill_sections.py [hunk 6]
```diff
@@ -32,51 +42,78 @@ def reference(reference_id: str) -> NoEscape:
...
-    section_p_text = pydash.get(obj=paragraph, path="text")
-    yield section_p_text
+
+    ref_ids = [item["refId"] for item in text_element if item.get("type") == "ref"]
+
+    x = []
+    for item in ref_ids:
+        x.append(reference(item))
+
+    return x
```
Comment (thenav56) [inline app/pdf_utils/fill_sections.py:77]: Maybe we can use a single iteration.
```suggestion
    return [
       reference(item["refId"])
        for item in text_element
        if item.get("type") == "ref"
    ]
```

File: app/pdf_utils/fill_sections.py [hunk 7]
```diff
@@ -32,51 +42,78 @@ def reference(reference_id: str) -> NoEscape:
...
-    for _indx, paragraph in enumerate(section_element["children"]):
+    for _indx, text_leaf in enumerate(section_element["children"]):
+
+        # check if the text_leaf has the path text
+        section_p_text = pydash.get(obj=text_leaf, path="text")
 
-        # check if the paragraph has the path text
-        # if so get text
-        section_p_text = pydash.get(obj=paragraph, path="text")
+        if len(section_element["children"]) > 1:
```
Comment (thenav56) [inline app/pdf_utils/fill_sections.py:89]: We can remove this as `for child in section_element["children"]:` will do the same thing with an empty array. This way we can have 1 less nested code.

File: app/pdf_utils/fill_sections.py [hunk 8]
```diff
@@ -32,51 +42,78 @@ def reference(reference_id: str) -> NoEscape:
...
 
-        #         # run functions in text_wrappers
-        #         section_p_text = command(section_p_text)
+            # if the option in text_leaf is true
+            if text_leaf.get(option):
+                # run functions in text_wrappers
+                section_p_text = wrap_text(text_leaf)
 
-    return section_p_text
+        return section_p_text
```
Comment (thenav56) [inline app/pdf_utils/fill_sections.py:119]: Hey @naomatheus, this section seems to be complex. Maybe we can simplify this?
We are iterating section_element["children"] twice.

Trying to simplify we can get something like this. Is this expected?
```python
    ref_ids: list = get_reference_ids(section_element["children"])

    for text_leaf in section_element["children"]:
        # check if the text_leaf has the path text
        section_p_text = pydash.get(obj=text_leaf, path="text")

        if ref_ids:  # If there are ref ids
            ids_ = " ".join(str(i) for i in ref_ids)
            # return reference text and bibliography refs
            return NoEscape(f"{section_p_text} {ids_}")

        # XXX: Only this is required right for wrapping text?
        section_p_text = wrap_text(text_leaf)

        # XXX: Asumming there are only 1 children?
        return section_p_text
```

**Review by thenav56 — COMMENTED** [review]
LGTM

Just maybe we can clean out commented code if not required. `Debug code` just to flag code used for debugging which also we might need to clean up.

There are a few suggestions as well.

**Pull request closed without merging**